### PR TITLE
lightning: make OpLevelOptional suppress the error of DoChecksum

### LINKED ISSUE #45382: lightning still fail at checksum stage when set `checksum = "optional"`
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

set `checksum = "optional"` in configuration file, and checksum meets error [tikv:9005]Region is unavailable

### 2. What did you expect to see? (Required)

lightning import task is passed

### 3. What did you see instead (Required)

lightning import task is failed

### PR DESCRIPTION
**Pull request by lyzx2001**
### What problem does this PR solve?

Issue Number: close #45382 

Problem Summary:

lightning still fail at checksum stage when set `checksum = "optional"`.

### What is changed and how it works?

Make `OpLevelOptional` suppress the error of `DoChecksum`.

Since the needed integration test is quite simple, I added it into existing integration test `lightning_routes`.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/importer/table_import.go [hunk 1]
```diff
@@ -1034,6 +1037,16 @@ func (tr *TableImporter) postProcess(
 
 			var remoteChecksum *local.RemoteChecksum
 			remoteChecksum, err = DoChecksum(ctx, tr.tableInfo)
+			failpoint.Inject("checksum-error", func() {
+				logutil.CL(ctx).Debug("failpoint checksum-error injected.")
```
Comment (lance6716) [inline br/pkg/lightning/importer/table_import.go:1041]: better to use logger in `tr.logger`

File: br/pkg/lightning/importer/table_import.go [hunk 2]
```diff
@@ -1034,6 +1037,16 @@ func (tr *TableImporter) postProcess(
...
+				err = status.Error(codes.Unknown, "Checksum meets error.")
```
Comment (lance6716) [inline br/pkg/lightning/importer/table_import.go:1042]: `remoteChecksum` should also set to nil. In order to mimic a function returning error, we should handle all the return values. Then you will notice a bug.

File: br/tests/lightning_routes/run.sh [hunk 3]
```diff
@@ -4,12 +4,17 @@
 
 set -eux
 
+echo "testing checksum-error..."
+export GO_FAILPOINTS="github.com/pingcap/tidb/br/pkg/lightning/importer/checksum-error=1*return(true)"
```
Comment (lance6716) [inline br/tests/lightning_routes/run.sh:8]:
```suggestion
export GO_FAILPOINTS="github.com/pingcap/tidb/br/pkg/lightning/importer/checksum-error=1*return()"
```

I remember this is OK

File: br/pkg/lightning/importer/table_import.go (hunk 2 above)
Comment (lyzx2001) [reply br/pkg/lightning/importer/table_import.go:1042]: get it, modified

File: br/pkg/lightning/importer/table_import.go [hunk 4]
```diff
@@ -1034,17 +1036,30 @@ func (tr *TableImporter) postProcess(
...
-			}
-			err = tr.compareChecksum(remoteChecksum, localChecksum)
-			// with post restore level 'optional', we will skip checksum error
+			failpoint.Inject("checksum-error", func() {
+				tr.logger.Info("failpoint checksum-error injected.")
+				remoteChecksum = nil
+				err = status.Error(codes.Unknown, "Checksum meets error.")
+			})
 			if rc.cfg.PostRestore.Checksum == config.OpLevelOptional {
 				if err != nil {
```
[the remark this answers is not in the document]
Comment (lyzx2001) [reply br/pkg/lightning/importer/table_import.go]: modified

**Review by lance6716 — APPROVED** [review]

**Comment by GMHDBJD** [comment]
can you also update it for importinto? https://github.com/pingcap/tidb/blob/67e8f902f32fa793f45808bd7468fb6fe4bc5249/disttask/importinto/subtask_executor.go#L122

**Comment by lyzx2001** [comment]
> can you also update it for importinto?
> 
> https://github.com/pingcap/tidb/blob/67e8f902f32fa793f45808bd7468fb6fe4bc5249/disttask/importinto/subtask_executor.go#L122

updated

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**